src/main/scala/dpla/ingestion3/harvesters/oai/OaiResponseProcessor.scala

```diff
@@ -60,20 +80,102 @@ object OaiResponseProcessor {
-    *         The error code if it exists otherwise empty string
+    * @param source OaiSource
+    *               The single-page response to a single OAI query
+    *
+    * @param setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]])
+    *                   A partial function indicating how sets should be filtered,
+    *                   ie. according to a whitelist or blacklist.
+    */
+  def getSets(source: OaiSource, setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]]): OaiResponse = {
+
+    // Parse all sets from XML.
+    def getSetsFromXml(xml: Node): Seq[OaiSet] = {
+      val xmlSets = xml \\ "OAI-PMH" \\ "set"
+
+      xmlSets.flatMap(set => set.headOption match {
+        case Some(node) => {
+          val id = getSetIdentifier(node)
+          Some(OaiSet(id, node.toString, source))
+        }
+        case _ => None
+      })
+    }
+
+    val text = source.text.getOrElse("")
+
+    getXml(text) match {
+      case Failure(e) => {
+        OaiError(e.toString, source)
+      }
+      case Success(xml) => {
+        val sets = getSetsFromXml(xml)
+        // Apply partial function to filter out any sets that do not match some criteria.
+        val filtered = setFilter(sets)
+        SetsPage(filtered)
+      }
+    }
+  }
+
+  /**
+    * Accepts a set from an OAI feed an returns the OAI identifier
+    *
+    * @param set Node
+    *            The original set from the OAI feed
+    * @return The local identifier
+    */
+  def getSetIdentifier(set: Node): String = {
+    (set \\ "setSpec").text
+  }
+
+  def getAllSets(page: OaiSource): OaiResponse = {
+    val setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]] = {
+      // No filtering required; return all sets.
+      case(sets) => sets
+    }
+    getSets(page, setFilter)
+  }
+
+  // Get all sets that belong to a given whitelist.
+  def getSetsByWhitelist(page: OaiSource, whitelist: Array[String]) = {
+    val setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]] = {
+      // Return only those sets with ids NOT in the blacklist.
+      case(sets) => sets.filter { set => whitelist.contains(set.id) }
+    }
+    getSets(page, setFilter)
+  }
+
+  // Get all sets except those belonging to a given blacklist.
+  def getSetsByBlacklist(page: OaiSource, blacklist: Array[String]) = {
+    val setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]] = {
+      // Return only those sets with ids in the whitelist.
+      case(sets) => sets.filterNot{ set => blacklist.contains(set.id) }
+    }
+    getSets(page, setFilter)
+  }
+
+  /**
+    * Parse an error message from an XML node.  Throw an excpetion if an error
+    * is found.
+    *
+    * @param xml: NodeSeq
+    *             The XML that may include an error message.
+    *
+    * @return Try[Unit]
+    *
+    * @throws OaiHarvesterException
     *
     */
-  @throws(classOf[OaiHarvesterException])
-  def getOaiErrorCode(xml: NodeSeq): Unit = {
+  def getOaiErrorCode(xml: NodeSeq): Try[Unit] = Try{
```

Comment:
This would probably be cleaner as an if, although I wonder if Codacy would complain! :)

```
if ((xml \\ "OAI-PMH" \\ "error").nonEmpty)  
  throw new OaiHarvesterException((xml \\ "OAI-PMH" \\ "error").text.trim)
```